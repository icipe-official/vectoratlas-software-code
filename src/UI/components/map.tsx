import React, { useRef, useEffect } from 'react';

import Map from 'ol/Map';
import View from 'ol/View';
import TileLayer from 'ol/layer/Tile';
import VectorTileLayer from 'ol/layer/VectorTile';
import VectorTileSource from 'ol/source/VectorTile';
import MVT from 'ol/format/MVT';
import {transform} from 'ol/proj';
import {Style, Fill, Stroke} from 'ol/style';
import XYZ from 'ol/source/XYZ';

import { useAppSelector } from '../state/hooks';

const defaultStyle = new Style({
  fill: new Fill({
    color: [0,0,0,0]
  }),
  stroke: new Stroke({
    color: 'white',
    width: 0.5
  })
});

export const MapWrapper= () => {
  const mapStyles = useAppSelector(state => state.config.map_styles);

  const layerStyles = Object.assign({}, ...mapStyles.layers.map((layer:any) => ({[layer.name]: new Style({
    fill: new Fill({
      color: layer.fillColor
    }),
    stroke: layer.strokeColor ? new Stroke({ 
      color: layer.strokeColor,
      width: layer.strokeWidth
    }): undefined,
    zIndex: layer.zIndex
  })})));

  const mapElement =useRef(null);

  useEffect(() => {
    const initialMap = new Map({
      target: 'mapDiv',
      layers: [
        new VectorTileLayer({
          source: new VectorTileSource({
            attributions:
              '&copy; OpenStreetMap contributors, Whoâ€™s On First, ' +
              'Natural Earth, and osmdata.openstreetmap.de',
            format: new MVT(),
            maxZoom: 5,
            url: '/data/world/{z}/{x}/{y}.pbf',
          }),
          style: (feature) => {
            const layerName = feature.get('layer');
            return layerStyles[layerName] ?? defaultStyle;
          },
        })
        ,
        new TileLayer({
          source: new XYZ({
            url: '/data/an_gambiae/{z}/{x}/{y}.png',
            maxZoom: 5,
          }),
          opacity: 1.0
        })
      ],
      view: new View({
        center: transform([20, -5], 'EPSG:4326', 'EPSG:3857'),
        zoom: 4
      })
    });

    return () => initialMap.setTarget(undefined);
  }, [layerStyles]);
  return (
    <div id='mapDiv' ref={mapElement} style={{height:'90vh', width: '99.3vw'}}></div>
  );

};
