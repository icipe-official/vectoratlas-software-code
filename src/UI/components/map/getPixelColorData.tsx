export const getPixelColorData = (rgba:any) => {
  return ((rgba[0] + rgba[1] + rgba[2])*0.1);
};
