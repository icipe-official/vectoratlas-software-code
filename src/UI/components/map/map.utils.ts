export function unpackOverlays(map_layers: any) {
  if (map_layers.length === 0) {
    return [[], []];
  }
  const worldLayer = map_layers.find((l: any) => l.name === 'world');
  const overlayList = map_layers.filter((l: any) => l.name !== 'world');
  const worldMapLayers = worldLayer.overlays.map((o: any) => ({
    ...o,
    sourceLayer: worldLayer.name,
    sourceType: worldLayer.sourceType,
  }));
  return [overlayList, worldMapLayers];
}

export function responseToGEOJSON(occurrenceData: any) {
  const geoJSONPoints = (occurrenceData || []).map((d: any) => ({
    type: 'Feature',
    geometry: d.site.location,
    properties: {
      name: d.site.name,
      year_start: d.year_start,
      n_all: d.sample.n_all,
      species: d.recorded_species.species.species,
      series: d.recorded_species.species.series,
    },
  }));
  const geoJSONFeatureCollection = {
    type: 'FeatureCollection',
    features: geoJSONPoints,
  };
  return JSON.stringify(geoJSONFeatureCollection);
}
