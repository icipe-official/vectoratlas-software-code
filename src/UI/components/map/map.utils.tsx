export const getPixelColorData = (rgba: any) => {
  return (rgba[0] + rgba[1] + rgba[2]) * 0.1;
};

export function pixelHoverInteraction(
  e: any,
  layer: any,
  getDataFunction: Function,
  targetHTML: any
) {
  if (e.dragging) {
    return;
  }
  const pixelData = layer.getData(e.pixel);
  targetHTML.innerText = pixelData
    ? getDataFunction(pixelData).toFixed(2)
    : '0.00';
}

export function responseToGEOJSON(graphqlLocationObject){
  var pointGeoJSONArray = [];
  for (let i = 0; i < graphqlLocationObject.data.allGeoData.length; i++) {
    const graphqlPointObject = graphqlLocationObject.data.allGeoData[i];
    const pointGeoJSON = 
    { type: 'Feature',
      geometry: {
        type: 'Point', 
        coordinates: [graphqlPointObject.longitude, graphqlPointObject.latitude]
      },
      properties: {
        arbData: graphqlPointObject.value
      }
    };
    pointGeoJSONArray.push(pointGeoJSON);
  }
  const geoJSONFeatureCollection = 
  {type: 'FeatureCollection',
    features: pointGeoJSONArray
  };
  return JSON.stringify(geoJSONFeatureCollection);
}
