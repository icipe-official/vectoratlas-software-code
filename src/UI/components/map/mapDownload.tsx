import { Button, Link, Paper } from '@mui/material'
import React from 'react'

export default function MapDownload(): JSX.Element {

  return (
    <>

    <div id="map" className='map'></div>

   <Link 
        id="export-png" 
        className="btn btn-outline-dark" 
        role="button"
        underline='hover'
        variant='overline'
        fontSize='20px'
        style={
            {   display: 'flex', 
                justifyContent: 'center' ,
                
                
             }
        }>
            <i className="fa fa-download"></i> DOWNLOAD MAP IMAGE</Link>
    <Link id="image-download" download="map.png"></Link>

    </>
  )
};

