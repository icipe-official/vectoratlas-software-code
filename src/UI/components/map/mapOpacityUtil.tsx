// Opacity Control Functionality:  
export function updateOpacity(layer:any, opacityInput:any, opacityOutput:any) {
  const opacity = parseFloat(opacityInput.value);
  layer.setOpacity(opacity);
  opacityOutput.innerText = opacity.toFixed(2);
}
