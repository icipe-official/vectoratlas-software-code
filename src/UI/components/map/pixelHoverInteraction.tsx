export function pixelHoverInteraction( e:any, layer:any ,getDataFunction:Function, targetHTML:any ) {
  if (e.dragging) {
    return;
  }
  const pixelData = layer.getData(e.pixel);
  targetHTML.innerText = pixelData ? getDataFunction(pixelData).toFixed(2) : '0.00';
}
