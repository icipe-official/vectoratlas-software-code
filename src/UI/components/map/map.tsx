import React, { useRef, useEffect, useState } from 'react';

import Box from '@mui/material/Box';

import Map from 'ol/Map';
import View from 'ol/View';
import TileLayer from 'ol/layer/Tile';
import VectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';
import VectorTileLayer from 'ol/layer/VectorTile';
import VectorTileSource from 'ol/source/VectorTile';
import MVT from 'ol/format/MVT';
import { transform } from 'ol/proj';
import { Circle, Style, Fill, Stroke } from 'ol/style';
import XYZ from 'ol/source/XYZ';
import GeoJSON from 'ol/format/GeoJSON';
import Text from 'ol/style/Text';
import 'ol/ol.css';

import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { responseToGEOJSON } from './utils/map.utils';
import { getOccurrenceData, getSpeciesList } from '../../state/mapSlice';
import DrawerMap from './layers/drawerMap';

const defaultStyle = new Style({
  fill: new Fill({
    color: [0, 0, 0, 0],
  }),
  stroke: new Stroke({
    color: 'white',
    width: 0.5,
  }),
});

export const MapWrapper = () => {
  const mapStyles = useAppSelector((state) => state.map.map_styles);
  const occurrenceData = useAppSelector((state) => state.map.occurrence_data);
  const layerVisibility = useAppSelector((state) => state.map.map_overlays);
  const mapOverlays = useAppSelector((state) => state.map.map_overlays);

  const overlaysList = mapOverlays.filter(
    (l: any) => l.sourceLayer !== 'world'
  );

  const seriesArray = useAppSelector((state) => state.map.species_list);

  const dispatch = useAppDispatch();

  const [map, setMap] = useState<Map | null>(null);

  useEffect(() => {
    console.log('getting data');
    dispatch(getOccurrenceData());
    dispatch(getSpeciesList());
  }, [dispatch]);

  const layerStyles = Object.assign(
    {},
    ...mapStyles.layers.map((layer: any) => ({
      [layer.name]: new Style({
        fill: new Fill({
          color: layerVisibility.find((l: any) => l.name === layer.name)
            ?.isVisible
            ? layer.fillColor
            : [0, 0, 0, 0],
        }),
        stroke: layer.strokeColor
          ? new Stroke({
              color: layerVisibility.find((l: any) => l.name === layer.name)
                ?.isVisible
                ? layer.strokeColor
                : [0, 0, 0, 0],
              width: layer.strokeWidth,
            })
          : undefined,
        zIndex: layer.zIndex,
      }),
    }))
  );

  const mapElement = useRef(null);

  useEffect(() => {
    function markStyle(n_all: number, seriesString: string) {
      return new Style({
        image: new Circle({
          radius: 15,
          fill: new Fill({
            color: seriesArray.find((s: any) => s.series === seriesString)
              ?.color ?? [0, 0, 0, 0.7],
          }),
          stroke: new Stroke({
            color: '0',
            width: 1,
          }),
        }),
        text: new Text({
          text: n_all !== null ? String(n_all) : '',
          scale: 1.4,
          fill: new Fill({
            color: '#fff',
          }),
          stroke: new Stroke({
            color: '0',
            width: 1,
          }),
        }),
      });
    }

    function buildRasterLayer(layer: any) {
      const layerXYZ = new XYZ({
        url: `/data/${layer.name}/{z}/{x}/{y}.png`,
        maxZoom: 5,
      });
      const tileLayer = new TileLayer({
        preload: Infinity,
        source: layerXYZ,
        opacity: 1.0,
        visible: layerVisibility.find((l: any) => l.name === layer.name)
          ?.isVisible,
      });
      tileLayer.set('name', layer.name);

      return tileLayer;
    }

    const pointLayer = new VectorLayer({
      source: new VectorSource({
        features: new GeoJSON().readFeatures(
          responseToGEOJSON(occurrenceData),
          {
            featureProjection: 'EPSG:3857',
          }
        ),
      }),
      style: (feature) => {
        return markStyle(feature.get('n_all'), feature.get('series'));
      },
    });
    pointLayer.set('occurrence-data', true);

    const baseMapLayer = new VectorTileLayer({
      preload: Infinity,
      source: new VectorTileSource({
        attributions: 'Made with Natural Earth. cc Vector Atlas',
        format: new MVT(),
        maxZoom: 5,
        url: '/data/world/{z}/{x}/{y}.pbf',
      }),
      style: (feature) => {
        const layerName = feature.get('layer');
        return layerStyles[layerName] ?? defaultStyle;
      },
    });
    baseMapLayer.set('base-map', true);

    // Passing in layers to generate map with overlays
    const initialMap = new Map({
      target: 'mapDiv',
      layers: [
        baseMapLayer,
        ...overlaysList.map((l: any) => buildRasterLayer(l)),
        pointLayer,
      ],
      view: new View({
        center: transform([20, -5], 'EPSG:4326', 'EPSG:3857'),
        zoom: 4,
      }),
    });


    setMap(initialMap);

    // Initialise map
    return () => initialMap.setTarget(undefined);
  }, [mapStyles, seriesArray]);

  useEffect(() => {
    const pointsLayer = map
      ?.getAllLayers()
      .find((l) => l.get('occurrence-data'));

    pointsLayer?.setSource(
      new VectorSource({
        features: new GeoJSON().readFeatures(
          responseToGEOJSON(occurrenceData),
          {
            featureProjection: 'EPSG:3857',
          }
        ),
      })
    );
  }, [map, occurrenceData]);

  useEffect(() => {
    const allLayers = map?.getAllLayers();
    allLayers?.forEach((l) => {
      const matchingLayer = layerVisibility.find(
        (v) => v.name === l.get('name')
      );
      if (matchingLayer) {
        l.setVisible(matchingLayer.isVisible);
      }
    });

    const baseMapLayer = map
      ?.getAllLayers()
      .find((l) => l.get('base-map')) as VectorTileLayer;
    baseMapLayer?.setStyle((feature) => {
      const layerName = feature.get('layer');
      return layerStyles[layerName] ?? defaultStyle;
    });
  }, [map, layerVisibility, mapStyles]);

  useEffect(() => {
    document.getElementById('export-png')?.addEventListener('click', function () {
      map?.once('rendercomplete', function () {
        const mapCanvas = document.createElement('canvas');
        const size = map.getSize();
        if (!size || size.length < 2) {
          return;
        }
        mapCanvas.width = size[0];
        mapCanvas.height = size[1];
        const mapContext = mapCanvas.getContext('2d');
        if (!mapContext) {
          return;
        }
        Array.prototype.forEach.call(
          map.getViewport().querySelectorAll('.ol-layer canvas, canvas.ol-layer'),
          function (canvas) {
            if (canvas.width > 0) {
              const opacity =
                canvas.parentNode.style.opacity || canvas.style.opacity;
              mapContext.globalAlpha = opacity === '' ? 1 : Number(opacity);
              let matrix;
              const transform = canvas.style.transform;

              if (transform) {
                matrix = transform
                  .match(/^matrix\(([^\(]*)\)$/)[1]
                  .split(',')
                  .map(Number);
              } else {
                matrix = [
                  parseFloat(canvas.style.width) / canvas.width,
                  0,
                  0,
                  parseFloat(canvas.style.height) / canvas.height,
                  0,
                  0,
                ];
              }
              CanvasRenderingContext2D.prototype.setTransform.apply(
                mapContext,
                matrix
              );
              const backgroundColor = canvas.parentNode.style.backgroundColor;
              if (backgroundColor) {
                mapContext.fillStyle = backgroundColor;
                mapContext.fillRect(0, 0, canvas.width, canvas.height);
              }

              mapContext.drawImage(canvas, 0, 0);
            }
          }
        );
        mapContext.globalAlpha = 1;
        mapContext.setTransform(1, 0, 0, 1, 0, 0);
        const link = document.getElementById(
          'image-download'
        ) as HTMLAnchorElement | null;

        if (link != null) {
          link.href = mapCanvas.toDataURL();
          link.click();
        }
      });

        if(map?.getRenderer()){
          map?.renderSync();
        }


    });
  }, [map])


  return (
    <Box sx={{ display: 'flex', flexGrow: 1 }}>
      <DrawerMap />
      <Box component="main" sx={{ flexGrow: 1 }}>
        <div
          id="mapDiv"
          ref={mapElement}
          style={{ height: 'calc(100vh - 230px)' }}
          data-testid="mapDiv"
        ></div>
      </Box>
    </Box>
  );
};
