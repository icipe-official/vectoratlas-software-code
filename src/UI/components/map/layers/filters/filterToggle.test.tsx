import React from 'react';
import { filterHandler, initialState } from "../../../../state/map/mapSlice";
import { fireEvent, render } from "../../../../test_config/render";
import FilterToggle from './filterToggle';

describe("FilterToggle", () => {
  let state;

  beforeEach(() => {
    state = { map: initialState() };
  })

  it('renders correctly for a given filter', () => {
    state.map.filterValues.country = ["Kenya", "Uganda", "Tanzania"]

    const { wrapper } = render(
      <FilterToggle 
        filterTitle={'Season'} 
        filterName="season"
        filterToggleType={'string'}
        filterOptionsArray={[
          { name: 'rainy', optionIcon: <div>Rainy icon</div> },
          { name: 'dry', optionIcon: <div>Dry icon</div> },
          { name: 'empty', optionIcon: <div>Empty icon</div> },
        ]}
      />,
      state
    );

    expect(wrapper.container).toMatchSnapshot();
  })

  it('dispatches the correct filter update when the value changes', async () => {
    state.map.filters.season.value = ["dry"]

    const { wrapper, store } = render(
      <FilterToggle 
        filterTitle={'Season'} 
        filterName="season"
        filterToggleType={'string'}
        filterOptionsArray={[
          { name: 'rainy', optionIcon: <div>Rainy icon</div> },
          { name: 'dry', optionIcon: <div>Dry icon</div> },
          { name: 'empty', optionIcon: <div>Empty icon</div> },
        ]}
      />,
      state
    );

    const rainyButton = await wrapper.getByText('rainy')
    fireEvent.click(rainyButton)

    expect(store.getActions()[0]).toEqual(filterHandler({
      filterName: 'season',
      filterOptions: ["dry", "rainy"]
    }))
  })

  it('dispatches the correct filter update when the value changes for boolean filter', async () => {
    state.map.filters.isAdult.value = ["empty"]

    const { wrapper, store } = render(
      <FilterToggle 
        filterTitle={'Adult'} 
        filterName="isAdult"
        filterToggleType={'boolean'}
        filterOptionsArray={[
          { name: 'true', optionIcon: <div>True icon</div> },
          { name: 'false', optionIcon: <div>False icon</div> },
          { name: 'empty', optionIcon: <div>Empty icon</div> },
        ]}
      />,
      state
    );

    const trueButton = await wrapper.getByText('true')
    fireEvent.click(trueButton)

    expect(store.getActions()[0]).toEqual(filterHandler({
      filterName: 'isAdult',
      filterOptions: ["empty", true]
    }))
  })
})