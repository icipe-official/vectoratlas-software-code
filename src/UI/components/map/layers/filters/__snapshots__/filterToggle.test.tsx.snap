// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterToggle renders correctly for a given filter 1`] = `
<div>
  <div
    style="display: flex; flex-direction: column; margin-top: 5px; width: 100%;"
  >
    <div
      style="display: flex; flex-direction: row; margin-top: 5px; width: 100%; justify-content: space-between;"
    >
      <p
        class="MuiTypography-root MuiTypography-inherit css-99b57d-MuiTypography-root"
      >
        Season
      </p>
    </div>
    <div
      class="MuiToggleButtonGroup-root css-i9wkih-MuiToggleButtonGroup-root"
      role="group"
    >
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17r3stk-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="rainy"
      >
        <div>
          Rainy icon
        </div>
        rainy
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17r3stk-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="dry"
      >
        <div>
          Dry icon
        </div>
        dry
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-pressed="false"
        class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeSmall MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-17r3stk-MuiButtonBase-root-MuiToggleButton-root"
        tabindex="0"
        type="button"
        value="empty"
      >
        <div>
          Empty icon
        </div>
        empty
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;
